hell/adventcalendar/2025/35/index.md

```diff
@@ -1,35 +1,84 @@
+    <figcaption>Example validation message on Firefox</figcaption>
+</figure>
+
+<figure style="margin: 4rem 0;">
+    <img src="./safari.jpg" alt="A screenshot of Safari.  The error message says 'Match the requested format: Only alphabetical and numerical characters are accepted'" style="border: 1px solid #ccc;">
+    <figcaption>Example validation message on Safari</figcaption>
+</figure>
+
+<figure style="margin: 4rem 0;">
+    <img src="./ios.jpeg" alt="A screenshot of Safari on iOS.  The error message says 'Match the requested format: Only alphabetical and numerical characters are accepted'" style="border: 1px solid #ccc;">
+    <figcaption>Example validation message on Safari on iOS</figcaption>
+</figure>
+
+> You can play with a live example at [this CodePen](https://codepen.io/wes_goulet/pen/emJjqKj).
+
+BTW, I just noticed that Chrome and Firefox say "Please" but Safari doesn't 😃
+
+## Styling
+
+You can't style the error popup, so if that's important to you then maybe you need to write your own error UI. A lot of times I don't mind leaning on browser UI when it's available (ie: most of the time I don't need my error messages in my website's overall brand/styling). Also, I think a lot of users have seen the browser's error UI before (from other sites that use native form validation), so there is some familiarity there for the user.
+
+## The Problem: Accessibility
+
+Unfortunately, native form validation isn't very accessible.
+
+I had assumed it was accessible, because most of the time when I lean on the browser to do something it handles accessibility much better than any userland code I would write. But after reading [this excellent post from Adrian Roselli](https://adrianroselli.com/2019/02/avoid-default-field-validation.html) (an a11y expert), I learned my assumption was wrong.
+
+Adrian's post lists some bugs opened against browsers to fix a11y issues. Some are marked as fixed, some are not. Here's hoping browser makers continue to improve the accessibility of native form validation.
+
+## Make it better with JS
+
+I originally wrote this post to point out that native form validation is good enough (lean on the browser!) and that's all you need. But after reading about the accessibility issues, I think the right answer is using native form validation as the foundation, and then adding a bit of JS to make it more accessible.
+
+How do we do that? This [post by Cloud Four](https://cloudfour.com/thinks/progressively-enhanced-form-validation-part-2-layering-in-javascript/) spells it out nicely, so I won't repeat it here. (If you mainly support evergreen browsers then I wouldn't worry too much about the first part "Removing invalid styles on page load for all browsers" since Chrome has shipped support for `:user-invalid` for [a couple years now](https://caniuse.com/wf-user-pseudos).)
```

Comment:
Can you please mention that a solution is the [Constraint Validation API](https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation) and briefly explain what it is and how it works. And then show a small snippet so that we at least get an idea how it looks like.

For example something like:

```js
const element = document.getElementById("mySelect");
console.log(element.checkValidity());
```